src/mips/psyqo/src/advancedpad.cpp

```diff
@@ -0,0 +1,305 @@
+    while (SIO::Stat & SIO::Status::STAT_RXRDY) {
+        SIO::Data;  // throwaway read
+    }
+}
+
+constexpr uint8_t psyqo::AdvancedPad::output_analog(uint8_t ticks) {
+    uint8_t data_out = 0xff;
+
+    switch (ticks) {}
+
+    return data_out;
+}
+
+constexpr uint8_t psyqo::AdvancedPad::output_default(uint8_t ticks) {
+    uint8_t data_out = 0x00;
+    switch (ticks) {
+        case 0:
+        case 2:
+            data_out = 0x01;
+            break;
+        case 1:
+            data_out = 0x42;
+            break;
+    }
+
+    return data_out;
+}
+
+constexpr uint8_t psyqo::AdvancedPad::output_multitap(uint8_t ticks) {
+    uint8_t data_out = 0x00;
+
+    switch (ticks) {
+        case 0:  // Pad select
+        case 2:  // Buffered transfer
+            data_out = 0x01;
+            break;
+
+        case 1:   // Initial read command
+        case 3:   // Read pad A
+        case 11:  // Read pad B
+        case 19:  // Read pad C
+        case 27:  // Read pad D
+            data_out = 0x42;
+            break;
+    }
+
+    switch (ticks) {
+        case 2:                // Burst speed for idhi
+        case 11:               // Burst speed for Pads B-D
+            SIO::Baud = 0x22;  // 1MHz
+            break;
+
+        case 3:                // Return to normal speed for Pad A
+            SIO::Baud = 0x88;  // 250kHz
+            break;
+    }
+
+    return data_out;
+}
+
+void psyqo::AdvancedPad::processChanges(Pad pad) {
+    bool padConnected = isPadConnected(pad);
+    bool wasConnected = m_connected[pad];
+    if (wasConnected && !padConnected) {
+        m_callback(Event{Event::PadDisconnected, pad});
+    } else if (!wasConnected && padConnected) {
+        m_callback(Event{Event::PadConnected, pad});
+    }
+    m_connected[pad] = padConnected;
+
+    uint32_t mask = 1;
+    uint32_t padData = m_padData[pad][1];
+    uint32_t buttons = m_buttons[pad];
+    for (int i = 0; i < 16; i++, mask <<= 1) {
+        bool buttonPressed = (padData & mask) == 0;
+        bool wasButtonPressed = (buttons & mask) == 0;
+        if (buttonPressed && !wasButtonPressed) {
+            m_callback(Event{Event::ButtonPressed, pad, Button(i)});
+        } else if (!buttonPressed && wasButtonPressed) {
+            m_callback(Event{Event::ButtonReleased, pad, Button(i)});
+        }
+    }
+    m_buttons[pad] = padData;
+}
+
+inline uint8_t psyqo::AdvancedPad::transceive(uint8_t data_out) {
+    SIO::Data = data_out;
+
+    // Wait for transceive to complete and data to populate FIFO
+    while (!(SIO::Stat & SIO::Status::STAT_RXRDY))
+        ;
+
+    // Pull data from FIFO
+    return SIO::Data;
+}
+
+void psyqo::AdvancedPad::readPad() {
+    uint8_t data_in, data_out;
+    uint8_t port_dev_type[2] = {PadType::None, PadType::None};
+
+    syscall_memset(m_padData, 0xff, sizeof(m_padData));
+
+    uint8_t *pad_data;
+    static constexpr unsigned pad_data_width = 8;
+
+    for (uint16_t port = 0; port < 2; port++) {
```

Comment:
It looks like this scans all ports? Which is different from what the .hh says, about alternating ports per frame. Also, `unsigned`.

Reply:
Huh, guessing the description in the header is a mix of leftover from simplepad and/or copilot hallucination, mixed with my own derp. Judging from simplepad's description, this is mentioned as a drawback, so I'll just nuke the comment about alternating frames unless this seems like desirable behavior. simplepad's description below:

 * @details This class is meant to be used as a singleton, probably in
 * the `Application` derived class. It is a simple thunk to the BIOS'
 * PAD interface, and has the same caveats, namely that it should be
 * initialized prior using the BIOS' memory card functions, and that
 * polling will alternate between the two pads at each frame when two
 * pads are connected, which can introduce input lags.